akka-docs/src/main/paradox/scala/guide/tutorial_5.md

```diff
@@ -0,0 +1,247 @@
+
+Java
+:   @@snip [DeviceGroupQueryTest.java]($code$/java/jdocs/tutorial_4/DeviceGroupQueryTest.java) { #query-test-stopped-later }
+
+The final case is when not all devices respond in time. To keep our test relatively fast, we will construct the
+`DeviceGroupQuery` actor with a smaller timeout:
+
+Scala
+:   @@snip [DeviceGroupQuerySpec.scala]($code$/scala/tutorial_4/DeviceGroupQuerySpec.scala) { #query-test-timeout }
+
+Java
+:   @@snip [DeviceGroupQueryTest.java]($code$/java/jdocs/tutorial_4/DeviceGroupQueryTest.java) { #query-test-timeout }
+
+Our query works as expected now, it is time to include this new functionality in the `DeviceGroup` actor now.
+
+## Adding query capability to the group
+
+Including the query feature in the group actor is fairly simple now. We did all the heavy lifting in the query actor
+itself, the group actor only needs to create it with the right initial parameters and nothing else.
+
+Scala
+:   @@snip [DeviceGroup.scala]($code$/scala/tutorial_4/DeviceGroup.scala) { #query-added }
+
+Java
+:   @@snip [DeviceGroup.java]($code$/java/jdocs/tutorial_4/DeviceGroup.java) { #query-added }
+
+It is probably worth restating what we said at the beginning of the chapter. By keeping the temporary state that is only relevant to the query itself in a separate actor we keep the group actor implementation very simple. It delegates
+everything to child actors and therefore does not have to keep state that is not relevant to its core business. Also, multiple queries can now run parallel to each other, in fact, as many as needed. In our case querying an individual device actor is a fast operation, but if this were not the case, for example, because the remote sensors need to be contacted over the network, this design would significantly improve throughput.
+
+We close this chapter by testing that everything works together. This test is just a variant of the previous ones, now exercising the group query feature:
+
+Scala
+:   @@snip [DeviceGroupSpec.scala]($code$/scala/tutorial_4/DeviceGroupSpec.scala) { #group-query-integration-test }
+
+Java
+:   @@snip [DeviceGroupTest.java]($code$/java/jdocs/tutorial_4/DeviceGroupTest.java) { #group-query-integration-test }
+
+## Summary
+In the context of the IoT system, this guide introduced the following concepts, among others. You can follow the links to review them if necessary:
+* [[The hierarchy of actors and their lifecycle|tutorial_1]]
+* [[The importance of designing messages for flexibility|tutorial_3.md]]
+* [[How to watch and stop actors, if necessary|tutorial_4.md#Keeping-track-of-the device-actors-in-the-group]]
+
+To continue your journey with Akka, we recommend:
+
+Reviewers: I wasn't sure where to send them next, you could provide a list of links to sections of the doc, or ebooks, or whatever. But since there are so many places they could go, it would be nice to provide at least a few of the logical "next" steps. :-)
```

Comment:
Good question. Perhaps indeed link to http://akka.io/get-involved/ and http://doc.akka.io/docs/akka/current/scala/additional/books.html ?